src/Microsoft.AspNet.Server.Kestrel/Http/SocketInput.cs

```diff
@@ -35,80 +56,172 @@ public ArraySegment<byte> Take(int count)
             return taken;
         }
 
-        public void Free()
+        public IncomingBuffer IncomingStart(int minimumSize)
         {
-            if (Buffer.Count == 0 && Buffer.Array.Length != 0)
+            lock (_sync)
             {
-                _memory.FreeByte(Buffer.Array);
-                Buffer = new ArraySegment<byte>(_memory.Empty, 0, 0);
+                if (_tail != null && minimumSize <= _tail.Data.Offset + _tail.Data.Count - _tail.End)
+                {
+                    _pinned = _tail;
+                    var data = new ArraySegment<byte>(_pinned.Data.Array, _pinned.End, _pinned.Data.Offset + _pinned.Data.Count - _pinned.End);
+                    var dataPtr = _pinned.Pin();
+                    return new IncomingBuffer
+                    {
+                        Data = data,
+                        DataPtr = dataPtr,
+                    };
+                }
             }
+
+            _pinned = _memory.Lease(minimumSize);
+            return new IncomingBuffer
+            {
+                Data = _pinned.Data,
+                DataPtr = _pinned.Pin()
```

Comment:
It's pinned only when the block wasn't provided as a slab segment. e.g. in the case you require more than ~4k linear bytes.